### ISSUE TITLE
Recommended recipe for entity embeddings

### ISSUE THREAD

**Issue by petulla** [flairNLP/flair#1591]
Hi

I'm trying to design a recipe to get 'entity embeddings' after the NER step. I'd like to use your word embeddings API, but I'm wondering how to optimize it. At the moment, my code look a bit like this:

```python
from flair.data import Sentence
from flair.models import SequenceTagger
tagger = SequenceTagger.load('ner')
from flair.embeddings import PooledFlairEmbeddings
# init embedding
pe = PooledFlairEmbeddings('news-forward', pooling='min')

t = tagger.predict
s = Sentence('This is a sentence about Bank of America. This is also a sentence about Apple computers. But it is also about the iPhone phone.')
t(s)

for entity in s.to_dict(tag_type='ner')['entities']:
  z = Sentence(entity['text'])
  print(z)
  pe.embed(z)
  for token in z:
    print(token)
    print(token.embedding)
  print('#~')
```

Output:

```python
Sentence: "Bank of America." - 3 Tokens
Token: 1 Bank
tensor([-0.0098,  0.0003,  0.0088,  ..., -0.0029, -0.0990,  0.0098],
       device='cuda:0')
Token: 2 of
tensor([ 1.1902e-03,  8.7077e-05,  2.1351e-02,  ..., -2.4563e-03,
         5.8907e-03,  4.3574e-03], device='cuda:0')
Token: 3 America.
tensor([-0.0012,  0.0016, -0.4050,  ..., -0.0006,  0.0058,  0.0172],
       device='cuda:0')
#~
Sentence: "Apple" - 1 Tokens
Token: 1 Apple
tensor([-0.0016,  0.0008, -0.1110,  ..., -0.0004, -0.0042,  0.0097],
       device='cuda:0')
#~
Sentence: "iPhone" - 1 Tokens
Token: 1 iPhone
tensor([-2.1312e-04, -9.8402e-06,  8.7166e-02,  ..., -2.6429e-04,
        -1.3806e-02,  2.8818e-02], device='cuda:0')
#~
```

The issue here is of course the nested loop and unpooled word embeddings. What I'd like instead is to get a single embedding back for each NER text span. I'm wondering if

a. You recommend an alternative to averaging/pooling this way
b. There's a better pattern here. I can make a comprehension that averages the tensors of course.. but I wanted to check if there's a simpler pattern.
c. Whether there's a specific way to fine-tune entity embeddings across the span, to get back a better word vector on an entity span by entity span basis rather than as word tokens.

**Comment by alanakbik** [comment]
Hi @petulla if you do it this way, you're embedding the spans, which means the sentence context is lost that some embeddings require. Better would be to embed the entire sentence first. Like this for instance: 

```python
from flair.data import Sentence
from flair.models import SequenceTagger

# init tagger
tagger = SequenceTagger.load('ner')

# init embedding
embeddings = FlairEmbeddings('news-forward')

# tag example sentence
sentence = Sentence('This is a sentence about Bank of America. This is also a sentence about Apple computers. But it is also about the iPhone phone.')
tagger.predict(sentence)

# embed entire sentence
embeddings.embed(sentence)

# iterate over entities
entities = sentence.get_spans(label_type='ner')
for entity in entities:

    # print
    print(entity)

    # iterate over each token in entity
    for token in entity.tokens:
        print(token)
        # print(token.embedding)
```

Averaging is probably your easiest option. You could also try using only the embedding of the first and last word of each span and concatenating them. 

Fine-tuning would require training data. But if you have it you could try fine-tuning a transformer embedding for instance.

**Comment by petulla** [comment]
Thanks @alanakbik. I'm wondering how I could fine-tune using the transformer embeddings not on subword or word embeddings but on the NER span itself. Does that even seem remotely doable? I'm thinking I'd need to alter the vocabulary to include multiple word tokens and really alter the architecture around that idea.. Wondering if you had any thoughts.

**Comment by alanakbik** [comment]
If I understand the idea correctly it would be difficult, since your vocabulary would explode. The reason most use subwords as "atoms" is so that there is a manageable vocabulary, like 50k subwords or so. If you use unique entity strings you would have a potentially unbounded vocabulary (the space of all possible names). 

Alternatively, you could take an existing transformer and fine-tune it on an NER task. Then extract the embeddigs at the subwords and average them.

**Comment by petulla** [comment]
I do plan to try both. I have one pipeline with HF's NER trained Electra etc.. Another with Flair. Although I don't disagree the vocabulary would be unbounded, it might be possible work with a bounded set and use subword averaging to handle tokens not in the vocabulary. I just wouldn't know how to set such a model up even if I had a vocabulary file (which I might be able to get..)

**Comment by petulla** [comment]
Realized I didn't really phrase the last comment as a question or in a way that it setup a need for feedback. But I'm just wondering if you had any other pointers or thoughts. I was thinking of how sense2vec captures phrases as a way to bound the multi-word embeddings. I just wouldn't really know how to train such a model.

**Comment by alanakbik** [comment]
I would probably try transformer embeddings without fine-tuning first, using the first and last subtoken, or by averaging over all tokens. For fine-tuning, you would have to find a task to fine-tune on that matches what you want to do with the NER embedding, but I have no immediate ideas for that. What do you want to use the NER embeddings for?

**Comment by petulla** [comment]
Simple answer is I'm training an embedding space based on entities. So I need entity embeddings as my input. To start, I'm averaging over all spans:

`torch.mean(torch.stack([token.embedding for token in entity.tokens]), dim=0).tolist()`

If any ideas come to mind let me know. I may drop one or two in here in the future if I need Flair's help.

**Comment by alanakbik** [comment]
Great, keep us posted!
